### PR TITLE
CGMES post processor

### LINKED ISSUE #686: CGMES post processor
* **Do you want to request a *feature* or report a *bug*?**
A feature. 
As CGMES allows additional profiles (like DL or GL) but also private extensions (RDF triples not in cim namespace), we need to way optionally and decoupled from core conversion to process this additional data and extend IIDM model with extensions.

* **What is the current behavior?**
The only way to do it in current code is to create an import post-processor. It works but there is several issues related to this:
  - import post processor are configured globally, so there is not way to only activate it on CGMES import
  - it is not possible in import parameters (properties), to select for a particular import which data extensions to process and which have to be skipped. 
  - In the importer post processor, we need to get the triple store which is done in an ugly way which classical post processor.

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**
N/A

* **What is the expected behavior?**
We need to support CGMES specific import post processor. Like for global post processor, it has to be discovered at runtime using service loader. CGMES import has to support a parameter to control active post processor.

```java
public interface CgmesImportPostProcessor {

    String getName();

    void process(Network network, TripleStore tripleStore);
}
```

* **What is the motivation / use case for changing the behavior?**

* **Please tell us about your environment:**
  - PowSyBl Version: 2.4.0-SNAPSHOT
  - OS Version: Linux

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, gitter, etc)

### PR DESCRIPTION
**Pull request by geofjamg**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
This pull-request fixes #686 

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**
We can define CGMES post processors

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No, just a new functionality added.

* **Other information**:

### CODE REVIEW

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java [hunk 1]
```diff
@@ -22,49 +15,79 @@
...
+    public static final Parameter POWSYBL_TRIPLESTORE_PARAMETER = new Parameter(POWSYBL_TRIPLESTORE, ParameterType.STRING, "The triplestore used during the import", TripleStoreFactory.defaultImplementation())
             .addAdditionalNames("powsyblTripleStore");
-    private static final Parameter STORE_CGMES_MODEL_AS_NETWORK_EXTENSION_PARAMETER = new Parameter(STORE_CGMES_MODEL_AS_NETWORK_EXTENSION, ParameterType.BOOLEAN, "Store the initial CGMES model as a network extension", Boolean.TRUE)
+    public static final Parameter STORE_CGMES_MODEL_AS_NETWORK_EXTENSION_PARAMETER = new Parameter(STORE_CGMES_MODEL_AS_NETWORK_EXTENSION, ParameterType.BOOLEAN, "Store the initial CGMES model as a network extension", Boolean.TRUE)
             .addAdditionalNames("storeCgmesModelAsNetworkExtension");
+    public static final Parameter POST_PROCESSORS_PARAMETER = new Parameter(POST_PROCESSORS, ParameterType.STRING_LIST, "Post processors", Collections.emptyList())
+            .addAdditionalNames("postProcessors");
+
+    private static final List<Parameter> PARAMETERS = Arrays.asList(CHANGE_SIGN_FOR_SHUNT_REACTIVE_POWER_FLOW_INITIAL_STATE_PARAMETER,
```
Comment (mathbagu) [inline cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java]: Should be an immutable list: Arrays.asList returns a fixed-size list, but values are modifiable using the set method

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java [hunk 2]
```diff
@@ -22,49 +15,79 @@
...
+                                                                    POWSYBL_TRIPLESTORE_PARAMETER,
+                                                                    STORE_CGMES_MODEL_AS_NETWORK_EXTENSION_PARAMETER,
+                                                                    POST_PROCESSORS_PARAMETER);
+
+    private final Map<String, CgmesImportPostProcessor> postProcessors;
 
     private final ParameterDefaultValueConfig defaultValueConfig;
 
-    public CgmesImport(PlatformConfig platformConfig) {
+    public CgmesImport(PlatformConfig platformConfig, List<CgmesImportPostProcessor> postProcessors) {
```
Comment (mathbagu) [inline cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java]: Should be a Set<CgmesImportPostProcessor>, but it probably break the API

File: cgmes/cgmes-model/src/main/java/com/powsybl/cgmes/model/CgmesModel.java [hunk 3]
```diff
@@ -22,6 +23,8 @@
  */
 public interface CgmesModel {
 
+    TripleStore tripleStore();
```
Comment (mathbagu) [inline cgmes/cgmes-model/src/main/java/com/powsybl/cgmes/model/CgmesModel.java]: getTripleStore?

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java (hunk 1 above)
Comment (geofjamg) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java]: Replaced by ImmutableList.of()

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java (hunk 2 above)
Comment (geofjamg) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java]: It directly comes from ServiceLoaderCache which return a List, but without any duplicated services.

File: cgmes/cgmes-model/src/main/java/com/powsybl/cgmes/model/CgmesModel.java (hunk 3 above)
Comment (geofjamg) [reply cgmes/cgmes-model/src/main/java/com/powsybl/cgmes/model/CgmesModel.java]: This methods was already present in implementation class and I didn't want to change it.
I added a missing @Override in impl class.

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java [hunk 4]
```diff
@@ -22,49 +15,79 @@
...
+    public static final Parameter POWSYBL_TRIPLESTORE_PARAMETER = new Parameter(POWSYBL_TRIPLESTORE, ParameterType.STRING, "The triplestore used during the import", TripleStoreFactory.defaultImplementation())
             .addAdditionalNames("powsyblTripleStore");
-    private static final Parameter STORE_CGMES_MODEL_AS_NETWORK_EXTENSION_PARAMETER = new Parameter(STORE_CGMES_MODEL_AS_NETWORK_EXTENSION, ParameterType.BOOLEAN, "Store the initial CGMES model as a network extension", Boolean.TRUE)
+    public static final Parameter STORE_CGMES_MODEL_AS_NETWORK_EXTENSION_PARAMETER = new Parameter(STORE_CGMES_MODEL_AS_NETWORK_EXTENSION, ParameterType.BOOLEAN, "Store the initial CGMES model as a network extension", Boolean.TRUE)
             .addAdditionalNames("storeCgmesModelAsNetworkExtension");
+    public static final Parameter POST_PROCESSORS_PARAMETER = new Parameter(POST_PROCESSORS, ParameterType.STRING_LIST, "Post processors", Collections.emptyList())
+            .addAdditionalNames("postProcessors");
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java]: Ok I removed the alias. I was thinking it is a just a simpler alias to long parameter name.

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java [hunk 5]
```diff
@@ -22,49 +15,79 @@
...
-    public CgmesImport(PlatformConfig platformConfig) {
+    public CgmesImport(PlatformConfig platformConfig, List<CgmesImportPostProcessor> postProcessors) {
         this.defaultValueConfig = new ParameterDefaultValueConfig(platformConfig);
+        this.postProcessors = Objects.requireNonNull(postProcessors).stream().collect(Collectors.toMap(CgmesImportPostProcessor::getName, e -> e));
     }
 
     public CgmesImport() {
-        this(PlatformConfig.defaultConfig());
+        this(PlatformConfig.defaultConfig(), new ServiceLoaderCache<>(CgmesImportPostProcessor.class).getServices());
+    }
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/CgmesImport.java]: Added

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/Conversion.java [hunk 6]
```diff
@@ -33,40 +33,19 @@
 public class Conversion {
 
     public Conversion(CgmesModel cgmes) {
-        this(cgmes, new Config());
+        this(cgmes, new Config(), Collections.emptyList());
     }
 
-    public Conversion(CgmesModel cgmes, Conversion.Config config) {
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/Conversion.java]: Added

**Pull request merged**